### PR TITLE
Issue 192

### PR DESCRIPTION
**Pull request by jamesros161**
This pull request is to resolve [https://github.com/BoldGrid/boldgrid-backup/issues/192](https://github.com/BoldGrid/boldgrid-backup/issues/192) . 

The issue arrises when attempting to import a database ( particularly during a migration ) that has views in addition to tables. If the user does not have the correct privileges, the archive restoration would succeed, but the views would not get imported.

This pull request addresses this by determining if the user has the correct privileges to restore the database, and if it does not, it will fail restoration and provide the user with the appropriate error message. 

Additionally, this will resolve issues with the "DEFINER" statement in a database that is being restored with views, when the old database user was named differently than the new database user.

### CODE REVIEW

File: admin/class-boldgrid-backup-admin-db-import.php [hunk 1]
```diff
@@ -166,8 +192,183 @@ public function import_lines( $lines ) {
...
+	 * @return array an array of results from the database query
+	 */
+	public function show_grants_query() {
+		$db           = new PDO( sprintf( 'mysql:host=%1$s;dbname=%2$s;', DB_HOST, DB_NAME ), DB_USER, DB_PASSWORD );
+		$db_statement = $db->query( 'SHOW GRANTS' );
+		return $db_statement->fetchAll();
+	}
+
+	/**
+	 * Execute Import
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-db-import.php]: Missing terminating full stop (period/dot).

File: admin/class-boldgrid-backup-admin-db-import.php [hunk 2]
```diff
@@ -166,8 +192,183 @@ public function import_lines( $lines ) {
...
+	 *
+	 * Executes Import MySql Query
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-db-import.php]: Missing terminating full stop (period/dot).

File: admin/class-boldgrid-backup-admin-db-import.php [hunk 3]
```diff
@@ -166,8 +192,183 @@ public function import_lines( $lines ) {
...
+	 *
+	 * @since SINCEVERSION
+	 *
+	 * @param PDO $db The PDO Object.
+	 * @param string $sql_line The line of sql to execute.
+	 *
+	 * @return int Number of affected rows
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-db-import.php]: Missing terminating full stop (period/dot).

File: admin/class-boldgrid-backup-admin-db-import.php [hunk 4]
```diff
@@ -166,8 +192,183 @@ public function import_lines( $lines ) {
...
+	 */
+	public function exec_import( PDO $db, $sql_line ) {
+		return $db->exec( $sql_line );
+	}
+
+	/**
+	 * Get a user's grants in the form of an array
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-db-import.php]: Missing terminating full stop (period/dot).

File: admin/class-boldgrid-backup-admin-db-import.php [hunk 5]
```diff
@@ -166,8 +192,183 @@ public function import_lines( $lines ) {
...
+	 * @since SINCEVERSION
+	 *
+	 * @param string $grants_string.
+	 * @return array An array of grants.
+	 */
+	public function get_grants_array( $grants_string ) {
+		if ( strpos( $grants_string, 'GRANT' ) === 0 ) {
+			$grants_string = substr( $grants_string, 6 );
+		}
+		if ( strpos( $grants_string, 'ON' ) ) {
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-db-import.php]: You may want to store the result of `strpos` in a variable for reuse.

File: admin/class-boldgrid-backup-admin-db-import.php [hunk 6]
```diff
@@ -166,8 +192,183 @@ public function import_lines( $lines ) {
...
+	 * Get a user's grants in the form of an array
+	 *
+	 * @since SINCEVERSION
+	 *
+	 * @param string $grants_string.
+	 * @return array An array of grants.
+	 */
+	public function get_grants_array( $grants_string ) {
+		if ( strpos( $grants_string, 'GRANT' ) === 0 ) {
+			$grants_string = substr( $grants_string, 6 );
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-db-import.php]: Use a comment to explain "magic number"; 6.

File: admin/class-boldgrid-backup-admin-core.php [hunk 7]
```diff
@@ -358,6 +358,14 @@ class Boldgrid_Backup_Admin_Core {
 	 */
 	public $db_omit;
 
+	/**
+	 * Database Restoration errro
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-core.php]: Missing terminating full stop (period/dot).

File: admin/class-boldgrid-backup-admin-db-import.php [hunk 8]
```diff
@@ -49,6 +48,23 @@ public function __construct( $core = false ) {
 		}
 	}
 
+	/**
+	 * Get Lines from file
+	 *
+	 * Gets an array of lines from a file
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-db-import.php]: Missing terminating full stop (period/dot).

File: admin/class-boldgrid-backup-admin-db-import.php [hunk 9]
```diff
@@ -49,6 +48,23 @@ public function __construct( $core = false ) {
 		}
 	}
 
+	/**
+	 * Get Lines from file
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-db-import.php]: Missing terminating full stop (period/dot).

File: admin/class-boldgrid-backup-admin-db-import.php [hunk 10]
```diff
@@ -49,6 +48,23 @@ public function __construct( $core = false ) {
...
+	 *
+	 * Gets an array of lines from a file
+	 *
+	 * @since SINCEVERSION
+	 *
+	 * @param string $file String ocntaining the path of the file.
+	 *
+	 * @return array An array of lines
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-db-import.php]: Missing terminating full stop (period/dot).

File: admin/class-boldgrid-backup-admin-db-import.php [hunk 11]
```diff
@@ -166,8 +192,183 @@ public function import_lines( $lines ) {
...
+	}
+
+	/**
+	 * Show Grants Query
+	 *
+	 * Queries the database for 'SHOW GRANTS'
+	 *
+	 * @since SINCEVERSION
+	 *
+	 * @return array an array of results from the database query
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-db-import.php]: Missing terminating full stop (period/dot).

File: admin/class-boldgrid-backup-admin-db-import.php [hunk 12]
```diff
@@ -166,8 +192,183 @@ public function import_lines( $lines ) {
...
+				return $this->get_grants_array( $result[0] );
+			}
+		}
+		return [];
+	}
+
+	/**
+	 * Show Grants Query
+	 *
+	 * Queries the database for 'SHOW GRANTS'
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-db-import.php]: Missing terminating full stop (period/dot).

File: tests/admin/test-class-boldgrid-backup-admin-db-import.php [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+
+		$this->expected_view_lines = [
+			'/*!50001 DROP VIEW IF EXISTS `test_view`*/;',
+			'/*!50001 CREATE ALGORITHM=UNDEFINED */',
+			'/*!50013 DEFINER=`' . DB_USER . '`@`localhost` SQL SECURITY DEFINER */',
+			'/*!50001 VIEW `test_view` AS select `wp_options`.`option_id` AS `option_id`,`wp_options`.`option_name` AS `option_name`,`wp_options`.`option_value` AS `option_value`,`wp_options`.`autoload` AS `autoload` from `wp_options` */;',
+		];
+	}
+
+	public function test_instance() {
```
Comment (cssjoe) [inline tests/admin/test-class-boldgrid-backup-admin-db-import.php]: This is not needed.  The `$core` class property in `Boldgrid_Backup_Admin_Db_Import` should remain private.  If really needed, then create a public getter method in the class.

File: admin/class-boldgrid-backup-admin-db-import.php [hunk 14]
```diff
@@ -22,10 +22,9 @@ class Boldgrid_Backup_Admin_Db_Import {
 	 * The core class object.
 	 *
 	 * @since 1.6.0
-	 * @access private
 	 * @var    Boldgrid_Backup_Admin_Core
 	 */
-	private $core;
+	public $core;
```
Comment (cssjoe) [inline admin/class-boldgrid-backup-admin-db-import.php]: This class property should be private.

File: admin/class-boldgrid-backup-admin-db-import.php (hunk 6 above)
Comment (jamesros161) [reply admin/class-boldgrid-backup-admin-db-import.php]: Instead, I stored the value 6, in a variable $expected_grants_string_start, and used that.

**Comment by cssjoe** [comment]
```There were 4 failures:

1) Test_Boldgrid_Backup_Admin_Core::test_restore_archive_file
Failed asserting that false is true.

/home/joec/public_html/wp-content/plugins/boldgrid-backup/tests/admin/test-class-boldgrid-backup-admin-core.php:194
/home/joec/public_html/wp-content/plugins/boldgrid-backup/tests/admin/test-class-boldgrid-backup-admin-core.php:323

2) Test_Boldgrid_Backup_Admin_Core::test_restore_cli
Failed asserting that false is true.

/home/joec/public_html/wp-content/plugins/boldgrid-backup/tests/admin/test-class-boldgrid-backup-admin-core.php:172
/home/joec/public_html/wp-content/plugins/boldgrid-backup/tests/admin/test-class-boldgrid-backup-admin-core.php:341

3) Test_Boldgrid_Backup_Admin_Db_Import::test_import_string
Failed asserting that 'The Database User does not have the necessary priviliges to restore this database.' matches expected 4.

/home/joec/public_html/wp-content/plugins/boldgrid-backup/tests/admin/test-class-boldgrid-backup-admin-db-import.php:127

4) Test_Boldgrid_Backup_Admin_Db_Import::test_fix_view_statements
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
+    0 => '/*!50001 DROP VIEW IF EXISTS ...ew`*/;'
+    1 => '/*!50001 CREATE ALGORITHM=UNDEFINED */'
+    2 => '/*!50013 DEFINER=`joec_wp_tes...NER */'
+    3 => '/*!50001 VIEW `test_view` AS ...s` */;'
 )

/home/joec/public_html/wp-content/plugins/boldgrid-backup/tests/admin/test-class-boldgrid-backup-admin-db-import.php:139

FAILURES!
Tests: 58, Assertions: 180, Failures: 4.```

File: admin/class-boldgrid-backup-admin-core.php (hunk 7 above)
Comment (cssjoe) [reply admin/class-boldgrid-backup-admin-core.php]: Still missing full stop, and typoed.

**Pull request merged**